zproject/backends.py

```diff
@@ -867,6 +888,62 @@ def sync_custom_profile_fields_from_ldap(
         except SyncUserException as e:
             raise ZulipLDAPException(str(e)) from e
 
+    def sync_groups_from_ldap(self, user_profile: UserProfile, ldap_user: _LDAPUser) -> None:
+        if not settings.LDAP_SYNC_GROUPS:
+            return
+
+        if user_profile.realm_id != settings.LDAP_GROUP_SYNC_REALM_ID:
+            return
+
+        try:
+            ldap_logger.debug("found groups for user: %s", str(user_profile))
+            ldap_group_name_set = set(ldap_user.group_names).intersection(set(settings.LDAP_GROUP_SYNC_LIST))
+
+            member_group_name_set = set(
+                [
+                    membership.user_group.name
+                    for membership in UserGroupMembership.objects.filter(
+                        user_group__realm=user_profile.realm,
+                        user_group__name__in=set(settings.LDAP_GROUP_SYNC_LIST),
+                        user_profile=user_profile,
+                    )
+                ]
+            )
```

Comment:
```suggestion
            member_group_name_set = set(UserGroupMembership.objects.filter(user_group__realm=user_profile.realm, user_group__name__in=set(settings.LDAP_GROUP_SYNC_LIST), user_profile=user_profile).distinct("user_group_id").values_list("user_group__name", flat=True)
                    ))
```
Something like this (didn't test the exact query, so might need some minor tweaks to work) should be more efficient, because in your snipper you're doing a `for` loop over all the group membering and in each iteration fetching `membership.group_name` in a separate query just to get the name.

Reply:
I re-wrote the part a bit. Adding and removing of groups is now done in bulk.

Reply:
added your performance fix.